tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,224 @@
+            // get component type
+            String type = components.get(component).getType();
+            if (type == null) {
+                ToolError.reportError(ErrorMessage.MEMBER_NOT_DECLARED,
+                        node.getLocation(), "component", component);
+            }
+            // check component of that type has port
+            ASTComponentTypeDecl typeDecl = componentTypes
+                    .get(type);
+            if (!typeDecl.getReqs().containsKey(port)
+                    && !typeDecl.getProvs().containsKey(port)) {
+                ToolError.reportError(ErrorMessage.MEMBER_NOT_DECLARED,
+                        node.getLocation(), "port", port);
+            }
+
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTPortDecl node, Object data) {
+        portdecls.put(node.getPort(), node);
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTEntryPointDecl node, Object data) {
+        entrypoints.put(node.getName(), node.getAction());
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTComponentDecl node, Object data) {
+        String name = node.getName();
+        String type = node.getType();
+        if (!componentTypes.containsKey(type)) {
+            // component type not declared
+            ToolError.reportError(ErrorMessage.ARCH_TYPE_NOT_DEFINED,
+                    node.getLocation(), "component", type);
+        }
+        if (components.putIfAbsent(name, node) != null) {
+            // duplicate members
+            ToolError.reportError(ErrorMessage.DUPLICATE_MEMBER_NAMES,
+                    node.getLocation(), name);
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTConnectorDecl node, Object data) {
+        String name = node.getName();
+        String type = node.getType();
+        if (!connectorTypes.contains(type)) {
+            // connector type not declared
+            ToolError.reportError(ErrorMessage.ARCH_TYPE_NOT_DEFINED,
+                    node.getLocation(), "connector", type);
+        }
+        if (connectors.putIfAbsent(name, type) != null) {
+            // duplicate members
+            ToolError.reportError(ErrorMessage.DUPLICATE_MEMBER_NAMES,
+                    node.getLocation(), name);
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTComponentTypeDecl node, Object data) {
+        if (componentTypes.containsKey(node.getTypeName())) {
+            // type of this name has already been declared
+            ToolError.reportError(ErrorMessage.DUPLICATE_TYPE_DEFINITIONS,
+                    node.getLocation(), node.getTypeName());
+        }
+        node.collectPorts();
+        if (!node.checkModule(state)) {
+            // module def not found HANDLED IN CHECKMODULE
+            return super.visit(node, data);
+        } else {
+            componentTypes.put(node.getTypeName(), node);
+        }
```

Comment:
I would change this if-else block as follows:

if (node.checkModule(state)) {
    componentTypes.put(node.getTypeName(), node);
}

Reply:
There are a number of errors that checkModule can throw depending on which checks it fails when the function is called on the ComponentTypeDecl node, so it would be difficult to throw the error directly from the DeclCheckVisitor. Would it be better to use a try/catch blow here when checkModule is invoked?

Reply:
Ah, in that case, we should be doing the following:

if (node.checkModule(state)) {
&nbsp;&nbsp;&nbsp;&nbsp;componentTypes.put(node.getTypeName(), node);
}

In checkModule(), you should just throw an exception whenever a check fails. That means you would never return false but I guess it's okay as it is in line with how error handling is done in the rest of Wyvern.

This applies to my comment on visit(ASTConnectorTypeDecl ...) as well.